test/GatewayUpgradeFlow.js

```diff
@@ -0,0 +1,180 @@
+            'function upgrade(address newImplementation, bytes32 newImplementationCodeHash, bytes calldata setupParams) external',
+        ]);
+        const newGatewayImplementation = await gatewayFactory.deploy(auth.address, tokenDeployer.address).then((d) => d.deployed());
+        const newGatewayImplementationCodeHash = await getBytecodeHash(newGatewayImplementation);
+        const setupParams = '0x';
+        const calldata = targetInterface.encodeFunctionData('upgrade', [
+            newGatewayImplementation.address,
+            newGatewayImplementationCodeHash,
+            setupParams,
+        ]);
+
+        const block = await ethers.provider.getBlock('latest');
+        const eta = block.timestamp + timeDelay;
+
+        const proposalHash = solidityKeccak256(['address', 'bytes', 'uint256'], [target, calldata, nativeValue]);
+
+        const payload = defaultAbiCoder.encode(
+            ['uint256', 'address', 'bytes', 'uint256', 'uint256'],
+            [commandID, target, calldata, nativeValue, eta],
+        );
+
+        await expect(interchainGovernance.executeProposalAction(governanceChain, governanceAddress, payload))
+            .to.emit(interchainGovernance, 'ProposalScheduled')
+            .withArgs(proposalHash, target, calldata, nativeValue, eta);
+    });
+
+    it('should execute an upgrade proposal on AxelarGateway', async () => {
+        const commandID = 0;
+        const target = gateway.address;
+        const nativeValue = 0;
+        const timeDelay = 12 * 60 * 60;
+
+        const targetInterface = new Interface([
+            'function upgrade(address newImplementation, bytes32 newImplementationCodeHash, bytes calldata setupParams) external',
+        ]);
+        const newGatewayImplementation = await gatewayFactory.deploy(auth.address, tokenDeployer.address).then((d) => d.deployed());
+        const newGatewayImplementationCodeHash = await getBytecodeHash(newGatewayImplementation);
+        const setupParams = '0x';
+        const calldata = targetInterface.encodeFunctionData('upgrade', [
+            newGatewayImplementation.address,
+            newGatewayImplementationCodeHash,
+            setupParams,
+        ]);
+
+        const block = await ethers.provider.getBlock('latest');
+        const eta = block.timestamp + timeDelay;
+
+        const proposalHash = solidityKeccak256(['address', 'bytes', 'uint256'], [target, calldata, nativeValue]);
+
+        const payload = defaultAbiCoder.encode(
+            ['uint256', 'address', 'bytes', 'uint256', 'uint256'],
+            [commandID, target, calldata, nativeValue, eta],
+        );
+
+        await expect(interchainGovernance.executeProposalAction(governanceChain, governanceAddress, payload))
+            .to.emit(interchainGovernance, 'ProposalScheduled')
+            .withArgs(proposalHash, target, calldata, nativeValue, eta);
+
+        await network.provider.send('evm_increaseTime', [timeDelay]);
```

Comment:
do a sleep for non hardhat networks

Reply:
used waitFor() function similar to InterchainGovernance tests